File: test1.py
# 不安全示範：直接把未經處理的使用者輸入拼到 SQL 會導致 SQL injection
def unsafe_login(conn, username: str, password: str) -> bool:
    # 直接拼接 user input -> 易受 SQL injection
    sql = "SELECT id FROM users WHERE username = '" + username + "' AND password = '" + password + "'"
    cursor = None
    try:
        cursor = conn.cursor()
        cursor.execute(sql)
        row = cursor.fetchone()
        return row is not None
    finally:
        if cursor:
            try:
                cursor.close()
            except Exception:
                pass

File: test1.java
import java.sql.*;

public class UnsafeAuthA {
    public boolean login(Connection conn, String username, String password) throws SQLException {
        // 直接拼接 user input -> 易受 SQL injection
        String sql = "SELECT id FROM users WHERE username = '" + username + "' AND password = '" + password + "'";
        Statement stmt = null;
        ResultSet rs = null;
        try {
            stmt = conn.createStatement();
            rs = stmt.executeQuery(sql);
            return rs.next();
        } finally {
            if (rs != null) try { rs.close(); } catch (Exception e) {}
            if (stmt != null) try { stmt.close(); } catch (Exception e) {}
        }
    }
}
